src/player.cpp

```diff
@@ -11489,8 +11489,15 @@ void player::read(int inventory_position)
         assign_activity( ACT_READ, time - moves, -1, inventory_position );
         // Never trigger studying when skimming the book.
         activity.values.push_back(0);
+        // pushes item_frm into 2nd element
+        activity.values.push_back(inv_from);
+        activity.values.push_back(item_fr_loc.x);
+        activity.values.push_back(item_fr_loc.y);
+        activity.values.push_back(item_fr_loc.z);
+        activity.values.push_back(item_fr_veh?-1:0);
+        activity.str_values.push_back(it->typeId());
```

Comment:
It might be a good approach to add some serialization function to `item_location` (could use the json classes). Something like this:
```C++
class item_location {
public:
    ...
    std::string serialize() const;
    static item_location deserialize( const std::string &data );
};

// Usage:
void start_activity() {
    item_location item_loc = ...;
    ...
    u.activity.str_values.push_back( item_loc.serialize() );
}
void finish_activity() {
    item_location item_loc = item_location::deserialize( u.activity.str_values[0] );
    if( item_loc.get_item() == nullptr ) {
        // item position might have changed.
    }
    ...
}
```

This encapsulates the data of the different types of item location. The user of that class doesn't need to know where the item actually is. `item_location` may add arbitrary properties of the item (type, damage, etc.) to the serialized string, and *check* those upon deserialization, returning a null-item when they don't match anymore. Again, that checking would be hidden inside `item_location` and can be extended anytime.